source/common/config/version_converter.h

```diff
@@ -32,11 +34,37 @@ class VersionConverter {
    */
   static void upgrade(const Protobuf::Message& prev_message, Protobuf::Message& next_message);
 
-  // Downgrade a message to the previous version. If no previous version exists,
-  // the given message is copied in the return value. This is not super
-  // efficient, most uses are expected to be tests and performance agnostic
-  // code.
-  static DowngradedMessagePtr downgrade(const Protobuf::Message& message);
+  /**
+   * Downgrade a message to the previous version. If no previous version exists,
+   * the given message is copied in the return value. This is not super
+   * efficient, most uses are expected to be tests and performance agnostic
+   * code.
+   *
+   * @param message message input.
+   * @return DynamicMessagePtr with the downgraded message (and associated
+   *         factory state).
+   */
+  static DynamicMessagePtr downgrade(const Protobuf::Message& message);
+
+  /**
+   * Reinterpret an Envoy internal API message at v3 based on a given API
+   * version. This will downgrade() to an earlier version or scrub the shadow
+   * deprecated fields in the existing one.
```

Comment:
sorry because I'm dense: the scrubbing here is for which case? Someone ingested v2, but we want to output v3, so we take the current converted v3, and just remove any of the hidden deprecated fields that got us here? But if we do that won't we lose context as the v3 message won't actually be accurate? I think I'm possibly just not understanding the use case.

Reply:
It's simpler than that. This is for an internal v3 proto (no ingestion), `DiscoveryRequest`. We have a deprecated field in it now, `build_version`, which has the weird mangling with `hidden_envoy_deprecated`. When we send it out on the wire, e.g. as a  v2 JSON REST message, it needs to still be around, but named with the original name, so we need to downgrade. When we send it out on the wire as v3, the field should be totally scrubbed, as it's not a valid v3 field.